src/Simulation/AutoSubstitution/RewriteStep.cs

```diff
@@ -0,0 +1,170 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.Quantum.QsCompiler.CsharpGeneration;
+using Microsoft.Quantum.QsCompiler.SyntaxTokens;
+using Microsoft.Quantum.QsCompiler.SyntaxTree;
+
+namespace Microsoft.Quantum.QsCompiler.AutoSubstitution
+{
+    /// <summary>
+    /// A Q# rewrite step for auto substitution
+    /// </summary>
+    ///
+    /// <para>
+    /// This rewrite step creates custom emulators for operations that have the
+    /// `SubstitutableOnTarget` attribute.  This attribute holds an alternative operation,
+    /// with the same signature, as its first argument, and a simulator, for which
+    /// the alternative operation should be used, as its second argument.
+    /// </para>
+    public class RewriteStep : IRewriteStep
+    {
+        public string Name => "AutoSubstitution";
+
+        // This rewrite step needs to be run before C# code generation
+        public int Priority => -2;
+
+        public IDictionary<string, string?> AssemblyConstants { get; } = new Dictionary<string, string?>();
+
+        public IEnumerable<IRewriteStep.Diagnostic> GeneratedDiagnostics => diagnostics;
+
+        public bool ImplementsPreconditionVerification => false;
+
+        public bool ImplementsTransformation => true;
+
+        public bool ImplementsPostconditionVerification => false;
+
+        public bool PostconditionVerification(QsCompilation compilation) => throw new System.NotImplementedException();
+
+        public bool PreconditionVerification(QsCompilation compilation) => throw new System.NotImplementedException();
+
+        public bool Transformation(QsCompilation compilation, [NotNullWhen(true)] out QsCompilation? transformed)
+        {
+            // we do not change the Q# syntax tree
+            transformed = compilation;
+
+            // global callables
+            var globalCallables = compilation.Namespaces.GlobalCallableResolutions();
+
+            // collect all callables that have an substitution attribute
+            var globals = globalCallables.Where(p => p.Value.Source.CodeFile.EndsWith(".qs"))
+                                         .Where(p => p.Value.Attributes.Any(HasSubstitutionAttribute));
+
+            if (!globals.Any())
+            {
+                diagnostics.Add(new IRewriteStep.Diagnostic {
```

Comment:
Formatting nit, but braces are typically on the next line for initializer syntax.
```suggestion
                diagnostics.Add(new IRewriteStep.Diagnostic
                {
```